pyvrp/cpp/Individual.cpp

```diff
@@ -170,18 +120,112 @@ Individual::Individual(ProblemData const &data, Routes routes)
     evaluate(data);
 }
 
-std::ostream &operator<<(std::ostream &out, Individual const &indiv)
+Individual::Route::Route(ProblemData const &data, Visits const visits)
+    : visits_(std::move(visits))
 {
-    auto const &routes = indiv.getRoutes();
+    if (visits_.empty())
+        return;
+
+    int time = data.depot().twEarly;
+    int prevClient = 0;
 
-    for (size_t rIdx = 0; rIdx != indiv.numRoutes(); ++rIdx)
+    for (size_t idx = 0; idx != size(); ++idx)
     {
-        out << "Route #" << rIdx + 1 << ":";  // route number
-        for (int cIdx : routes[rIdx])
-            out << " " << cIdx;  // client index
-        out << '\n';
+        distance_ += data.dist(prevClient, visits_[idx]);
+        duration_ += data.duration(prevClient, visits_[idx]);
+        demand_ += data.client(visits_[idx]).demand;
+        service_ += data.client(visits_[idx]).serviceDuration;
+
+        time += data.client(prevClient).serviceDuration
+                + data.duration(prevClient, visits_[idx]);
+
+        if (time < data.client(visits_[idx]).twEarly)  // add wait duration
+        {
+            wait_ += data.client(visits_[idx]).twEarly - time;
+            time = data.client(visits_[idx]).twEarly;
+        }
+
+        if (time > data.client(visits_[idx]).twLate)  // add time warp
+        {
+            timeWarp_ += time - data.client(visits_[idx]).twLate;
+            time = data.client(visits_[idx]).twLate;
+        }
+
+        prevClient = visits_[idx];
     }
 
+    Client const last = visits_.back();  // lLast client has depot as successor
+    distance_ += data.dist(last, 0);
+    duration_ += data.duration(last, 0);
+    time += data.client(last).serviceDuration + data.duration(last, 0);
+    timeWarp_ += std::max(time - data.depot().twLate, 0);  // depot closing tw
+    excessLoad_ = data.vehicleCapacity() < demand_
+                      ? demand_ - data.vehicleCapacity()
+                      : 0;
+}
```

Comment:
Doesn't have to be part of this PR but given that you're touching it now I would like to make the suggestion of using the TimeWindowSegment class to do the computations. That would eliminate the redundant logic of wait time/time warp computation and additionally we would also get the earliest/latest start times of the routes which is quite useful. What do you think of that (I can make it a separate issue)?

Reply:
Yeah that could work! It's probably best to handle this separately through a new issue. When we start working on that issue it could also be a good time to have another look at how TWS is currently used (and see if we can maybe get rid of its default constructor, since that thing makes no sense).

Reply:
Can you open an issue for this?